jolt-core/src/main/java/com/bazaarvoice/jolt/templatr/function/Math.java

```diff
@@ -18,151 +18,316 @@
 
 import com.bazaarvoice.jolt.common.Optional;
 
+import java.util.ArrayList;
+import java.util.List;
+
 @SuppressWarnings( "unused" )
 public class Math {
 
-    public static final class MaxOf implements Function {
-        @Override
-        public Optional<Object> apply( final Object... args ) {
-            Integer maxInt = Integer.MIN_VALUE;
-            Double maxDouble = -(Double.MIN_VALUE);
-            Long maxLong = Long.MIN_VALUE;
+    public static Optional<Number> MaxOf( Object... args ) {
+        if(args.length == 0) {
+            return Optional.empty();
+        }
 
-            if(args.length == 0) {
-                return Optional.empty();
+        Integer maxInt = Integer.MIN_VALUE;
+        Double maxDouble = -(Double.MAX_VALUE);
+        Long maxLong = Long.MIN_VALUE;
+
+        for(Object arg: args) {
+            if(arg instanceof Integer) {
+                maxInt = java.lang.Math.max( maxInt, (Integer) arg );
+            }
+            else if(arg instanceof Double) {
+                maxDouble = java.lang.Math.max( maxDouble, (Double) arg );
             }
+            else if(arg instanceof Long) {
+                maxLong = java.lang.Math.max(maxLong, (Long) arg);
+            }
+        }
+        if(maxInt == Integer.MIN_VALUE && maxDouble == ( - Double.MAX_VALUE ) && maxLong == Long.MIN_VALUE) {
+            return Optional.empty();
+        }
+        // explicit if else to avoid autoboxing
+        if(maxInt.longValue() >= maxDouble.longValue() && maxInt.longValue() >= maxLong) {
+            return Optional.<Number>of(maxInt);
+        }
+        else if(maxLong >= maxDouble.longValue() && maxLong >= maxInt.longValue()) {
```

Comment:
Optimizaion, in the "else if"  you don't have to check the Long vs the Int, cause you already know that that the Int is not the biggest thing, or else it would have been returned in the first "if" clause.